tests/test_edgeql_insert.py

```diff
@@ -1188,6 +1188,42 @@ async def test_edgeql_insert_for_17(self):
             ],
         )
 
+    async def test_edgeql_insert_for_bad_01(self):
+        with self.assertRaisesRegex(
+            edgedb.errors.QueryError,
+            "cannot reference correlated set",
+        ):
+            await self.con.execute("""
+                WITH MODULE test
+                SELECT (Person,
+                        (FOR x in {Person} UNION (
+                             INSERT Note {name := x.name})));
+            """)
+
+    async def test_edgeql_insert_for_bad_02(self):
+        with self.assertRaisesRegex(
+            edgedb.errors.QueryError,
+            "cannot reference correlated set",
+        ):
+            await self.con.execute("""
+                WITH MODULE test
+                SELECT (Person,
+                        (FOR x in {Person} UNION (
+                             SELECT (INSERT Note {name := x.name}))));
+            """)
+
+    async def test_edgeql_insert_for_bad_03(self):
```

Comment:
Let's add a test with a deeper `INSERT` placement to ensure the recursive nature of DML search (thinking of possible future refactors).